File: GFG POTD/022-September-2025/024-design-minmax-queue.java
// https://www.geeksforgeeks.org/problems/design-minmax-queue/1/




import java.util.*;

// TC : O(1) for all methods
// SC : O(N) 
class SpecialQueue {
    private Queue<Integer> q;
    private Deque<Integer> minQ;
    private Deque<Integer> maxQ;
    
    public SpecialQueue() {
        q = new LinkedList<>();
        minQ = new LinkedList<>();
        maxQ = new LinkedList<>();
    }

    public void enqueue(int x) {
        q.offer(x);
        
        while (!minQ.isEmpty() && minQ.peekLast() > x) minQ.pollLast();
        minQ.offerLast(x);
        
        while (!maxQ.isEmpty() && maxQ.peekLast() < x) maxQ.pollLast();
        maxQ.offerLast(x);
    }

    public void dequeue() {
        if (q.isEmpty()) return;
        int front = q.poll();
        
        if (!minQ.isEmpty() && front == minQ.peekFirst()) minQ.pollFirst();
        if (!maxQ.isEmpty() && front == maxQ.peekFirst()) maxQ.pollFirst();
    }

    public int getFront() {
        return q.isEmpty() ? -1 : q.peek();
    }

    public int getMin() {
        return minQ.isEmpty() ? -1 : minQ.peekFirst();
    }

    public int getMax() {
        return maxQ.isEmpty() ? -1 : maxQ.peekFirst();
    }
}

File: GFG POTD/022-September-2025/024-design-minmax-queue.cpp
// https://www.geeksforgeeks.org/problems/design-minmax-queue/1/




#include <bits/stdc++.h>
using namespace std;

class SpecialQueue {
  private:
    queue<int> q;
    deque<int> minQ, maxQ;
  public:

    void enqueue(int x) {
        q.push(x);
        
        // Maintain MinQ Increasing
        while (!minQ.empty() && minQ.back() > x) minQ.pop_back();
        minQ.push_back(x);
        
        // Maintain MaxQ Decreasing
        while (!maxQ.empty() && maxQ.back() < x) maxQ.pop_back();
        maxQ.push_back(x);
    }

    void dequeue() {
        if (q.empty()) return;
        int front = q.front();
        q.pop();
        
        // if the front is at the minQ or at maxQ pop it
        if (!minQ.empty() && front == minQ.front()) minQ.pop_front();
        if (!maxQ.empty() && front == maxQ.front()) maxQ.pop_front();

    }

    int getFront() {
        if (!q.empty()) return q.front();
        return -1;
    }

    int getMin() {
        if (!minQ.empty()) return minQ.front();
        return -1;
    }

    int getMax() {
        if (!maxQ.empty()) return maxQ.front();
        return -1;
    }
};